### ISSUE TITLE
View ops on fake tensors can dispatch `detach`es to backend kernels

### ISSUE THREAD

**Issue by charlie-wt** [pytorch/pytorch#104194]
### 🐛 Describe the bug

When calling an op on a fake tensor, a new fake tensor is made for the result via a call to `detach` ([src](https://github.com/pytorch/pytorch/blob/4e204ff87b1fbbe2f1820704e45559a6f448ce17/torch/csrc/autograd/python_variable.cpp#L587)).

When the op called is a view op (& I think it also needs to be one without a meta Python decomposition), the `detach` is dispatched to the backend kernel (eg. `CPU`), instead of to `Meta`.

This usually isn't an issue as `detach` is declared only `CompositExplicitAutograd`, but for out-of-tree backends that _do_ intercept `detach`, they'll end up getting an op dispatched with a `Meta` tensor.

# Reproducer

```python
import torch

t = torch.randn(5, 5)
with torch._subclasses.FakeTensorMode() as mode:
    ft = mode.from_tensor(t)
    res = ft.view(25)
    print(res)
```

Running this with `TORCH_SHOW_DISPATCH_TRACE=1` gives:

```
[call] op=[aten::randn], key=[BackendSelect]
  [redispatch] op=[aten::randn], key=[CPU]
   [call] op=[aten::empty.memory_format], key=[BackendSelect]
    [redispatch] op=[aten::empty.memory_format], key=[CPU]
   [call] op=[aten::normal_], key=[CPU]
 [call] op=[aten::empty_strided], key=[BackendSelect]
  [redispatch] op=[aten::empty_strided], key=[Meta]
 [call] op=[aten::detach], key=[AutogradMeta]
  [redispatch] op=[aten::detach], key=[ADInplaceOrView]
   [redispatch] op=[aten::detach], key=[Meta]
 [call] op=[aten::view], key=[PythonTLSSnapshot]
  [redispatchBoxed] op=[aten::view], key=[AutogradCPU]
   [redispatch] op=[aten::view], key=[ADInplaceOrView]
    [redispatch] op=[aten::view], key=[Python]
     [callBoxed] op=[aten::view], key=[Meta]
     [call] op=[aten::detach], key=[CPU]    <<<<<===== The issue
FakeTensor(..., size=(25,))
```

Running in a debugger, in `THPVariable_make_subclass` (called from `FakeTensor.__new__`), printing `toString(r.tensor(1).key_set())` gives `DispatchKeySet(CPU, ADInplaceOrView, AutogradCPU, AutocastCPU)`.

### Versions

Collecting environment information...
PyTorch version: 2.1.0a0+git4ff3108
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.25.0
[pip3] torch==2.1.0a0+git9593b3a
[conda] numpy                     1.25.0                   pypi_0    pypi
[conda] torch                     2.1.0a0+git9593b3a          pypi_0    pypi

**Comment by mikaylagawarecki** [comment]
cc @eellison

**Comment by charlie-wt** [comment]
just been taking another look at this, & the source of the issue is [here](https://github.com/pytorch/pytorch/blob/cbcd55104504609e7b42d208310514a9e8afb2c0/torch/_subclasses/fake_tensor.py#L985), where `FakeTensor`s are explicitly made with the dispatch keys of their fake device instead of `'meta'`. this was introduced in #82449, as a workaround for the benefit of autocast with `FakeTensor`s.

the pr & comments seem to imply that `__torch_dispatch__` should kick in an add `Meta` as a preferred backend compoment before ops are sent to a backend, hiding any issues. this isn't happening because the `detach` comes from within `FakeTensor.__torch_dispatch__` itself, outside of the `in_kernel_invocation_manager` that includes the `Meta` key—it comes from the call to `wrap_meta_outputs_with_default_device_logic` ([here](https://github.com/pytorch/pytorch/blob/cbcd55104504609e7b42d208310514a9e8afb2c0/torch/_subclasses/fake_tensor.py#L1532)). if i put that call in the `in_kernel_invocation_manager` too, the results seem to forget their fake device and become fully meta.
